keeper/reconcile.go

```diff
@@ -0,0 +1,129 @@
+			// TODO: Not sure if we should handle this here, perhaps just return nil?
+			// Starting of the initial period should be done by management process? Keep for now.
+			return k.VaultInterestDetails.Set(ctx, vault.GetAddress(), types.VaultInterestDetails{
+				PeriodStart: now,
+			})
+		}
+		return fmt.Errorf("failed to get vault interest details: %w", err)
+	}
+
+	if now <= interestDetails.PeriodStart {
+		return nil
+	}
+
+	duration := now - interestDetails.PeriodStart
+
+	reserves := k.BankKeeper.GetBalance(ctx, vault.GetAddress(), vault.UnderlyingAssets[0])
+	principal := k.BankKeeper.GetBalance(ctx, markertypes.MustGetMarkerAddress(vault.ShareDenom), vault.UnderlyingAssets[0])
+
+	interestEarned, err := interest.CalculateInterestEarned(principal, vault.InterestRate, duration)
+	if err != nil {
+		return fmt.Errorf("failed to calculate interest: %w", err)
+	}
+
+	if interestEarned.IsPositive() {
+		if reserves.Amount.LT(interestEarned) {
+			return fmt.Errorf("insufficient reserves to pay interest")
+		}
+
+		if err := k.BankKeeper.SendCoins(markertypes.WithBypass(ctx),
+			vault.GetAddress(),
+			markertypes.MustGetMarkerAddress(vault.ShareDenom),
+			sdk.NewCoins(sdk.NewCoin(vault.UnderlyingAssets[0], interestEarned)),
+		); err != nil {
+			return fmt.Errorf("failed to pay interest: %w", err)
+		}
+	} else if interestEarned.IsNegative() {
+		owed := interestEarned.Abs()
+		from := markertypes.MustGetMarkerAddress(vault.ShareDenom)
+		to := vault.GetAddress()
+
+		balance := k.BankKeeper.GetBalance(ctx, from, vault.UnderlyingAssets[0])
+		if balance.Amount.LT(owed) {
+			return fmt.Errorf("insufficient marker balance to reclaim negative interest")
+		}
+		if err := k.BankKeeper.SendCoins(markertypes.WithBypass(ctx),
+			from,
+			to,
+			sdk.NewCoins(sdk.NewCoin(vault.UnderlyingAssets[0], owed)),
+		); err != nil {
+			return fmt.Errorf("failed to reclaim negative interest: %w", err)
+		}
+	}
+
+	principalAfter := k.BankKeeper.GetBalance(ctx, markertypes.MustGetMarkerAddress(vault.ShareDenom), vault.UnderlyingAssets[0])
+
+	k.emitEvent(ctx, types.NewEventVaultReconcile(
+		vault.GetAddress().String(),
+		principal,
+		principalAfter,
+		vault.InterestRate,
+		duration,
+		interestEarned,
+	))
+
+	return k.VaultInterestDetails.Set(ctx, vault.GetAddress(), types.VaultInterestDetails{
+		PeriodStart: now,
+	})
+}
+
+// EstimateVaultTotalAssets returns the estimated total value of the vault's assets,
+// including any interest that would have accrued since the last interest period start.
+// This is used to simulate the value of earned interest as if a reconciliation had occurred
+// at the current block time, without modifying state or transferring funds.
+//
+// If no interest rate is set or if the vault has not yet begun accruing interest,
+// the original total asset amount is returned unmodified.
+func (k Keeper) EstimateVaultTotalAssets(ctx sdk.Context, vault *types.VaultAccount, totalAssets sdk.Coin) (sdkmath.Int, error) {
```

Comment:
I feel like we should have an explicit end date in the context here so that if there was a situation where this was queried but we hadn't closed out the period yet for what ever reason we would cap the amount of estimated interest instead of letting this pretend an unbounded period of time could have elapsed.  Perhaps our Vault-Interest module will be able to use a maximum timeout period/renewal period and clip the duration against that here ... that would also consolidate the configuration around the timeout process I discussed with @nullpointer0x00